File: data-structure/advanced/Easy Addition.c
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>

#define MOD 1000000007LL

// ---------------- Fast IO ----------------
typedef struct {
    unsigned char *buf;
    size_t idx, size;
} FastIn;

static FastIn IN;

static void init_fastin() {
    IN.buf = (unsigned char*)malloc(1 << 20);
    IN.idx = 0;
    IN.size = fread(IN.buf, 1, 1 << 20, stdin);
}

static inline int refill() {
    IN.idx = 0;
    IN.size = fread(IN.buf, 1, 1 << 20, stdin);
    return IN.size > 0;
}

static inline int read_int(int *out) {
    int c;
    do {
        if (IN.idx >= IN.size && !refill()) return 0;
        c = IN.buf[IN.idx++];
    } while (c <= ' ');

    int sign = 1;
    if (c == '-') { sign = -1; 
        if (IN.idx >= IN.size && !refill()) return 0;
        c = IN.buf[IN.idx++];
    }
    long long val = 0;
    while (c > ' ') {
        val = val * 10 + (c - '0');
        if (IN.idx >= IN.size && !refill()) break;
        c = IN.buf[IN.idx++];
    }
    *out = (int)(val * sign);
    return 1;
}

// ---------------- Modular ----------------
static inline long long mod_add(long long a, long long b){ a+=b; if(a>=MOD) a-=MOD; if(a<0) a+=MOD; return a; }
static inline long long mod_mul(long long a, long long b){ return (a*b)%MOD; }

static long long mod_pow(long long a, long long e){
    long long r=1%MOD;
    a%=MOD;
    while(e){
        if(e&1) r=mod_mul(r,a);
        a=mod_mul(a,a);
        e>>=1;
    }
    return r;
}
static inline long long mod_inv(long long a){ return mod_pow(a, MOD-2); }

// ---------------- Graph ----------------
typedef struct Edge { int to, nxt; } Edge;
static int *head;
static Edge *edges;
static int ecnt;

static void add_edge(int u,int v){
    edges[ecnt]=(Edge){v, head[u]};
    head[u]=ecnt++;
}

// ---------------- HLD + LCA ----------------
static int N;
static int *parent_, *depth_, *heavy, *sz_, *head_, *pos_, *invpos;
static int curpos;

static int LOG;
static int **up;

static void build_hld(int root){
    // iterative DFS to get parent/depth/order
    int *stack = (int*)malloc(sizeof(int)*N);
    int *order = (int*)malloc(sizeof(int)*N);
    int sp=0, ord=0;

    stack[sp++]=root;
    parent_[root]=0;
    depth_[root]=0;

    while(sp){
        int u=stack[--sp];
        order[ord++]=u;
        for(int e=head[u]; e!=-1; e=edges[e].nxt){
            int v=edges[e].to;
            if(v==parent_[u]) continue;
            parent_[v]=u;
            depth_[v]=depth_[u]+1;
            stack[sp++]=v;
        }
    }

    // sizes + heavy
    for(int i=ord-1;i>=0;i--){
        int u=order[i];
        sz_[u]=1;
        int best=0;
        heavy[u]=0;
        for(int e=head[u]; e!=-1; e=edges[e].nxt){
            int v=edges[e].to;
            if(v==parent_[u]) continue;
            sz_[u]+=sz_[v];
            if(sz_[v]>best){
                best=sz_[v];
                heavy[u]=v;
            }
        }
    }

    // decompose
    curpos=0;
    int *st_u = (int*)malloc(sizeof(int)*N);
    int *st_h = (int*)malloc(sizeof(int)*N);
    int sp2=0;
    st_u[sp2]=root; st_h[sp2]=root; sp2++;

    while(sp2){
        int u=st_u[--sp2];
        int h=st_h[sp2];

        int x=u;
        while(x){
            head_[x]=h;
            pos_[x]=++curpos;
            invpos[curpos]=x;

            for(int e=head[x]; e!=-1; e=edges[e].nxt){
                int v=edges[e].to;
                if(v==parent_[x] || v==heavy[x]) continue;
                st_u[sp2]=v; st_h[sp2]=v; sp2++;
            }
            x=heavy[x];
        }
    }

    free(stack); free(order);
    free(st_u); free(st_h);
}

static void build_lca(){
    LOG = 0;
    while((1<<LOG) <= N) LOG++;
    up = (int**)malloc(sizeof(int*)*LOG);
    for(int k=0;k<LOG;k++) up[k]=(int*)calloc(N+1,sizeof(int));

    for(int i=1;i<=N;i++) up[0][i]=parent_[i];
    for(int k=1;k<LOG;k++){
        for(int i=1;i<=N;i++){
            int mid=up[k-1][i];
            up[k][i]= mid? up[k-1][mid] : 0;
        }
    }
}

static int lca(int a,int b){
    if(depth_[a]<depth_[b]){ int t=a;a=b;b=t; }
    int diff=depth_[a]-depth_[b];
    for(int k=0; diff; k++){
        if(diff&1) a=up[k][a];
        diff>>=1;
    }
    if(a==b) return a;
    for(int k=LOG-1;k>=0;k--){
        if(up[k][a]!=up[k][b]){
            a=up[k][a];
            b=up[k][b];
        }
    }
    return parent_[a];
}

// ---------------- Segment Tree ----------------
//
// Supports range add of sequence: (p0+p1*i+p2*i^2)*q^i over i=0..len-1
// q is either R or invR.
// We store lazy coefficients separately for q=R and q=invR.
//
static long long *segSum;
static long long *lr0,*lr1,*lr2;
static long long *li0,*li1,*li2;

static long long *powR,*powI;
static long long *S0R,*S1R,*S2R;
static long long *S0I,*S1I,*S2I;

static inline void shift_poly(long long p0,long long p1,long long p2,long long off,
                              long long *o0,long long *o1,long long *o2){
    off%=MOD;
    long long off2 = mod_mul(off, off);
    *o0 = (p0 + mod_mul(p1, off) + mod_mul(p2, off2)) % MOD;
    *o1 = (p1 + mod_mul((2*p2)%MOD, off)) % MOD;
    *o2 = p2 % MOD;
}

static inline void add_to_node(int idx, long long p0,long long p1,long long p2, int useR, int len){
    long long add;
    if(useR){
        lr0[idx]=mod_add(lr0[idx], p0);
        lr1[idx]=mod_add(lr1[idx], p1);
        lr2[idx]=mod_add(lr2[idx], p2);
        add = ( mod_mul(p0,S0R[len]) + mod_mul(p1,S1R[len]) + mod_mul(p2,S2R[len]) ) % MOD;
    }else{
        li0[idx]=mod_add(li0[idx], p0);
        li1[idx]=mod_add(li1[idx], p1);
        li2[idx]=mod_add(li2[idx], p2);
        add = ( mod_mul(p0,S0I[len]) + mod_mul(p1,S1I[len]) + mod_mul(p2,S2I[len]) ) % MOD;
    }
    segSum[idx]=mod_add(segSum[idx], add);
}

static void push(int idx,int l,int r){
    if(l==r){
        lr0[idx]=lr1[idx]=lr2[idx]=0;
        li0[idx]=li1[idx]=li2[idx]=0;
        return;
    }
    int mid=(l+r)>>1;
    int left_len = mid-l+1;
    int right_len = r-mid;

    // push R-lazy
    long long p0=lr0[idx], p1=lr1[idx], p2=lr2[idx];
    if(p0||p1||p2){
        add_to_node(idx<<1, p0,p1,p2, 1, left_len);

        long long sp0,sp1,sp2;
        shift_poly(p0,p1,p2,left_len,&sp0,&sp1,&sp2);
        long long mul = powR[left_len];
        sp0=mod_mul(sp0,mul);
        sp1=mod_mul(sp1,mul);
        sp2=mod_mul(sp2,mul);
        add_to_node(idx<<1|1, sp0,sp1,sp2, 1, right_len);

        lr0[idx]=lr1[idx]=lr2[idx]=0;
    }

    // push invR-lazy
    p0=li0[idx]; p1=li1[idx]; p2=li2[idx];
    if(p0||p1||p2){
        add_to_node(idx<<1, p0,p1,p2, 0, left_len);

        long long sp0,sp1,sp2;
        shift_poly(p0,p1,p2,left_len,&sp0,&sp1,&sp2);
        long long mul = powI[left_len];
        sp0=mod_mul(sp0,mul);
        sp1=mod_mul(sp1,mul);
        sp2=mod_mul(sp2,mul);
        add_to_node(idx<<1|1, sp0,sp1,sp2, 0, right_len);

        li0[idx]=li1[idx]=li2[idx]=0;
    }
}

static void range_add_seq(int idx,int l,int r,int L,int R, int useR,
                          long long p0,long long p1,long long p2){
    if(R<l || r<L) return;
    if(L<=l && r<=R){
        int off = l - L;
        long long sp0,sp1,sp2;
        shift_poly(p0,p1,p2,off,&sp0,&sp1,&sp2);
        long long mul = useR ? powR[off] : powI[off];
        sp0=mod_mul(sp0,mul);
        sp1=mod_mul(sp1,mul);
        sp2=mod_mul(sp2,mul);
        add_to_node(idx, sp0,sp1,sp2, useR, r-l+1);
        return;
    }
    push(idx,l,r);
    int mid=(l+r)>>1;
    range_add_seq(idx<<1,l,mid,L,R,useR,p0,p1,p2);
    range_add_seq(idx<<1|1,mid+1,r,L,R,useR,p0,p1,p2);
    segSum[idx]=mod_add(segSum[idx<<1], segSum[idx<<1|1]);
}

static long long range_sum(int idx,int l,int r,int L,int R){
    if(R<l || r<L) return 0;
    if(L<=l && r<=R) return segSum[idx];
    push(idx,l,r);
    int mid=(l+r)>>1;
    return mod_add(range_sum(idx<<1,l,mid,L,R),
                   range_sum(idx<<1|1,mid+1,r,L,R));
}

// ---------------- Apply one HLD segment update ----------------
static long long Rbase, invR;

static void apply_segment(int useR, int Lpos, int Rpos, long long d0,
                          long long c0,long long c1,long long c2){
    d0%=MOD;
    long long d02 = mod_mul(d0,d0);
    long long base = mod_pow(Rbase, (long long)d0); // R^d0

    if(useR){
        // z=d0+i
        long long P2 = c2%MOD;
        long long P1 = (c1 + mod_mul(2*c2%MOD, d0)) % MOD;
        long long P0 = (c0 + mod_mul(c1,d0) + mod_mul(c2,d02)) % MOD;

        long long p0=mod_mul(P0,base), p1=mod_mul(P1,base), p2=mod_mul(P2,base);
        range_add_seq(1,1,N,Lpos,Rpos,1,p0,p1,p2);
    }else{
        // z=d0-i, exponent uses invR^i
        long long P2 = c2%MOD;
        long long P1 = (-c1 - mod_mul(2*c2%MOD, d0)) % MOD; if(P1<0) P1+=MOD;
        long long P0 = (c0 + mod_mul(c1,d0) + mod_mul(c2,d02)) % MOD;

        long long p0=mod_mul(P0,base), p1=mod_mul(P1,base), p2=mod_mul(P2,base);
        range_add_seq(1,1,N,Lpos,Rpos,0,p0,p1,p2);
    }
}

// update path A->B
static void path_update(int A,int B, long long a1,long long d1,long long a2,long long d2){
    a1%=MOD; d1%=MOD; a2%=MOD; d2%=MOD;
    long long c2 = mod_mul(d1,d2);
    long long c1 = (mod_mul(a1,d2) + mod_mul(a2,d1)) % MOD;
    long long c0 = mod_mul(a1,a2);

    int L = lca(A,B);
    int distAL = depth_[A] - depth_[L];

    // A -> L (inclusive), z decreases => use invR, chain direction top->bottom
    int u=A;
    while(head_[u]!=head_[L]){
        int h=head_[u];
        int Lp=pos_[h], Rp=pos_[u];
        long long d0 = (long long)(depth_[A] - depth_[h]);
        apply_segment(0, Lp, Rp, d0, c0,c1,c2);
        u=parent_[h];
    }
    // same head
    apply_segment(0, pos_[L], pos_[u], (long long)(depth_[A]-depth_[L]), c0,c1,c2);

    // L -> B (exclude L), z increases => use R
    // collect segments from top->bottom
    int v=B;
    // store segments in arrays (since O(logN) count)
    int segL[64], segR[64], sc=0;

    while(head_[v]!=head_[L]){
        int h=head_[v];
        segL[sc]=pos_[h];
        segR[sc]=pos_[v];
        sc++;
        v=parent_[h];
    }
    if(v!=L){
        segL[sc]=pos_[L]+1;
        segR[sc]=pos_[v];
        sc++;
    }

    // apply in reverse? These segs are already top->bottom on each chain,
    // and the overall order from L downward is the reverse of how we collected (from B up).
    // So apply from sc-1 downto 0.
    for(int i=sc-1;i>=0;i--){
        int lpos=segL[i], rpos=segR[i];
        int topnode = invpos[lpos];
        long long d0 = (long long)distAL + (long long)(depth_[topnode] - depth_[L]);
        apply_segment(1, lpos, rpos, d0, c0,c1,c2);
    }
}

static long long path_sum(int a,int b){
    long long res=0;
    while(head_[a]!=head_[b]){
        if(depth_[head_[a]] < depth_[head_[b]]){
            int t=a;a=b;b=t;
        }
        res = mod_add(res, range_sum(1,1,N, pos_[head_[a]], pos_[a]));
        a = parent_[head_[a]];
    }
    if(depth_[a] > depth_[b]){
        int t=a;a=b;b=t;
    }
    res = mod_add(res, range_sum(1,1,N, pos_[a], pos_[b]));
    return res;
}

// ---------------- Main ----------------
int main(){
    init_fastin();

    int Rtmp;
    read_int(&N);
    read_int(&Rtmp);
    Rbase = (long long)Rtmp % MOD;
    invR = mod_inv(Rbase);

    head = (int*)malloc((N+1)*sizeof(int));
    head_ = head;
    head = NULL;

    head_ = (int*)malloc((N+1)*sizeof(int));
    head = head_; // alias

    // graph
    head = (int*)malloc((N+1)*sizeof(int));
    for(int i=1;i<=N;i++) head[i]=-1;
    edges = (Edge*)malloc(sizeof(Edge)*2*(N-1));
    ecnt=0;

    for(int i=0;i<N-1;i++){
        int x,y;
        read_int(&x); read_int(&y);
        add_edge(x,y);
        add_edge(y,x);
    }

    parent_ = (int*)calloc(N+1,sizeof(int));
    depth_  = (int*)calloc(N+1,sizeof(int));
    heavy   = (int*)calloc(N+1,sizeof(int));
    sz_     = (int*)calloc(N+1,sizeof(int));
    head_   = (int*)calloc(N+1,sizeof(int));
    pos_    = (int*)calloc(N+1,sizeof(int));
    invpos  = (int*)calloc(N+1,sizeof(int));

    build_hld(1);
    build_lca();

    // segment tree arrays
    int SZ = 4*N + 5;
    segSum = (long long*)calloc(SZ,sizeof(long long));
    lr0 = (long long*)calloc(SZ,sizeof(long long));
    lr1 = (long long*)calloc(SZ,sizeof(long long));
    lr2 = (long long*)calloc(SZ,sizeof(long long));
    li0 = (long long*)calloc(SZ,sizeof(long long));
    li1 = (long long*)calloc(SZ,sizeof(long long));
    li2 = (long long*)calloc(SZ,sizeof(long long));

    // precompute pow and sums up to N
    powR = (long long*)malloc((N+1)*sizeof(long long));
    powI = (long long*)malloc((N+1)*sizeof(long long));
    powR[0]=1; powI[0]=1;
    for(int i=1;i<=N;i++){
        powR[i]=mod_mul(powR[i-1], Rbase);
        powI[i]=mod_mul(powI[i-1], invR);
    }

    S0R = (long long*)calloc(N+1,sizeof(long long));
    S1R = (long long*)calloc(N+1,sizeof(long long));
    S2R = (long long*)calloc(N+1,sizeof(long long));
    S0I = (long long*)calloc(N+1,sizeof(long long));
    S1I = (long long*)calloc(N+1,sizeof(long long));
    S2I = (long long*)calloc(N+1,sizeof(long long));

    for(int i=0;i<N;i++){
        long long qiR = powR[i];
        long long qiI = powI[i];
        S0R[i+1]=mod_add(S0R[i], qiR);
        S1R[i+1]=mod_add(S1R[i], mod_mul((long long)i, qiR));
        S2R[i+1]=mod_add(S2R[i], mod_mul(mod_mul((long long)i,(long long)i), qiR));

        S0I[i+1]=mod_add(S0I[i], qiI);
        S1I[i+1]=mod_add(S1I[i], mod_mul((long long)i, qiI));
        S2I[i+1]=mod_add(S2I[i], mod_mul(mod_mul((long long)i,(long long)i), qiI));
    }

    int U,Q;
    read_int(&U); read_int(&Q);

    // apply all updates
    for(int k=0;k<U;k++){
        int a1,d1,a2,d2,A,B;
        read_int(&a1); read_int(&d1); read_int(&a2); read_int(&d2); read_int(&A); read_int(&B);
        path_update(A,B, (long long)a1, (long long)d1, (long long)a2, (long long)d2);
    }

    // answer queries
    // output buffer
    char *out = (char*)malloc((size_t)Q * 24);
    size_t out_len=0;
    for(int k=0;k<Q;k++){
        int i,j;
        read_int(&i); read_int(&j);
        long long ans = path_sum(i,j) % MOD;
        out_len += (size_t)sprintf(out + out_len, "%lld\n", ans);
    }
    fwrite(out, 1, out_len, stdout);

    return 0;
}

File: data-structure/advanced/Easy Addition.py
import sys
sys.setrecursionlimit(1_000_000)
input = sys.stdin.readline

MOD = 1_000_000_007

def modinv(a):
    return pow(a, MOD - 2, MOD)

# ---------- Segment Tree supporting adds of (p0+p1*i+p2*i^2)*q^i ----------
class SegTree:
    __slots__ = ("n", "sum", "lr0", "lr1", "lr2", "li0", "li1", "li2",
                 "S0R","S1R","S2R","S0I","S1I","S2I","powR","powI")
    def __init__(self, n, R):
        self.n = n
        size = 4 * n + 5
        self.sum = [0] * size

        # lazy for base q=R
        self.lr0 = [0] * size
        self.lr1 = [0] * size
        self.lr2 = [0] * size
        # lazy for base q=invR
        self.li0 = [0] * size
        self.li1 = [0] * size
        self.li2 = [0] * size

        invR = modinv(R)

        # precompute powers up to n (max segment length)
        self.powR = [1] * (n + 1)
        self.powI = [1] * (n + 1)
        for i in range(1, n + 1):
            self.powR[i] = (self.powR[i-1] * R) % MOD
            self.powI[i] = (self.powI[i-1] * invR) % MOD

        # precompute S0,S1,S2 for q=R and q=invR:
        # S0[len]=sum q^i, S1[len]=sum i q^i, S2[len]=sum i^2 q^i for i=0..len-1
        def build_S(powq):
            S0 = [0] * (n + 1)
            S1 = [0] * (n + 1)
            S2 = [0] * (n + 1)
            for i in range(n):
                qi = powq[i]
                S0[i+1] = (S0[i] + qi) % MOD
                S1[i+1] = (S1[i] + i * qi) % MOD
                S2[i+1] = (S2[i] + (i*i % MOD) * qi) % MOD
            return S0, S1, S2

        self.S0R, self.S1R, self.S2R = build_S(self.powR)
        self.S0I, self.S1I, self.S2I = build_S(self.powI)

    def _add_to_node(self, idx, p0, p1, p2, useR, length):
        # add sum of (p0+p1*i+p2*i^2)*q^i over i=0..length-1
        if useR:
            S0, S1, S2 = self.S0R, self.S1R, self.S2R
            self.lr0[idx] = (self.lr0[idx] + p0) % MOD
            self.lr1[idx] = (self.lr1[idx] + p1) % MOD
            self.lr2[idx] = (self.lr2[idx] + p2) % MOD
        else:
            S0, S1, S2 = self.S0I, self.S1I, self.S2I
            self.li0[idx] = (self.li0[idx] + p0) % MOD
            self.li1[idx] = (self.li1[idx] + p1) % MOD
            self.li2[idx] = (self.li2[idx] + p2) % MOD

        add = (p0 * S0[length] + p1 * S1[length] + p2 * S2[length]) % MOD
        self.sum[idx] = (self.sum[idx] + add) % MOD

    @staticmethod
    def _shift_poly(p0, p1, p2, off):
        # poly(i+off) = p0' + p1'*i + p2*i^2
        offm = off % MOD
        off2 = (offm * offm) % MOD
        np0 = (p0 + p1 * offm + p2 * off2) % MOD
        np1 = (p1 + (2 * p2 % MOD) * offm) % MOD
        return np0, np1, p2

    def _push(self, idx, l, r):
        if l == r:
            # clear lazies
            self.lr0[idx]=self.lr1[idx]=self.lr2[idx]=0
            self.li0[idx]=self.li1[idx]=self.li2[idx]=0
            return
        mid = (l + r) >> 1
        left_len = mid - l + 1
        right_len = r - mid

        # push q=R lazy
        p0, p1, p2 = self.lr0[idx], self.lr1[idx], self.lr2[idx]
        if p0 or p1 or p2:
            # left child: same
            self._add_to_node(idx<<1, p0, p1, p2, True, left_len)
            # right child: shift by left_len and multiply by R^left_len
            sp0, sp1, sp2 = self._shift_poly(p0, p1, p2, left_len)
            mul = self.powR[left_len]
            sp0 = sp0 * mul % MOD
            sp1 = sp1 * mul % MOD
            sp2 = sp2 * mul % MOD
            self._add_to_node(idx<<1|1, sp0, sp1, sp2, True, right_len)
            self.lr0[idx]=self.lr1[idx]=self.lr2[idx]=0

        # push q=invR lazy
        p0, p1, p2 = self.li0[idx], self.li1[idx], self.li2[idx]
        if p0 or p1 or p2:
            self._add_to_node(idx<<1, p0, p1, p2, False, left_len)
            sp0, sp1, sp2 = self._shift_poly(p0, p1, p2, left_len)
            mul = self.powI[left_len]
            sp0 = sp0 * mul % MOD
            sp1 = sp1 * mul % MOD
            sp2 = sp2 * mul % MOD
            self._add_to_node(idx<<1|1, sp0, sp1, sp2, False, right_len)
            self.li0[idx]=self.li1[idx]=self.li2[idx]=0

    def range_add_seq(self, q_is_R, L, R, p0, p1, p2):
        # add (p0+p1*i+p2*i^2)*q^i on interval positions L..R, where i counts from 0 at L
        def rec(idx, l, r):
            if R < l or r < L:
                return
            if L <= l and r <= R:
                # need to shift by (l-L) because local i starts at 0 for node segment
                off = l - L
                sp0, sp1, sp2 = self._shift_poly(p0, p1, p2, off)
                mul = (self.powR[off] if q_is_R else self.powI[off])
                sp0 = sp0 * mul % MOD
                sp1 = sp1 * mul % MOD
                sp2 = sp2 * mul % MOD
                self._add_to_node(idx, sp0, sp1, sp2, q_is_R, r - l + 1)
                return
            self._push(idx, l, r)
            mid = (l + r) >> 1
            rec(idx<<1, l, mid)
            rec(idx<<1|1, mid+1, r)
            self.sum[idx] = (self.sum[idx<<1] + self.sum[idx<<1|1]) % MOD
        rec(1, 1, self.n)

    def range_sum(self, L, R):
        def rec(idx, l, r):
            if R < l or r < L:
                return 0
            if L <= l and r <= R:
                return self.sum[idx]
            self._push(idx, l, r)
            mid = (l + r) >> 1
            return (rec(idx<<1, l, mid) + rec(idx<<1|1, mid+1, r)) % MOD
        return rec(1, 1, self.n)

# ---------- HLD ----------
def solve():
    N, Rbase = map(int, input().split())
    g = [[] for _ in range(N+1)]
    for _ in range(N-1):
        x, y = map(int, input().split())
        g[x].append(y)
        g[y].append(x)

    parent = [0]*(N+1)
    depth  = [0]*(N+1)
    heavy  = [0]*(N+1)
    size   = [0]*(N+1)

    # iterative DFS for parent/depth/order
    order = [1]
    parent[1] = 0
    depth[1] = 0
    st = [1]
    while st:
        u = st.pop()
        order.append(u)
        for v in g[u]:
            if v == parent[u]: 
                continue
            parent[v] = u
            depth[v] = depth[u] + 1
            st.append(v)

    # compute subtree sizes & heavy child
    for u in reversed(order[1:]):
        size[u] = 1
        mx = 0
        for v in g[u]:
            if v == parent[u]: 
                continue
            size[u] += size[v]
            if size[v] > mx:
                mx = size[v]
                heavy[u] = v

    head = [0]*(N+1)
    pos  = [0]*(N+1)
    inv  = [0]*(N+1)
    cur = 0

    # decompose
    stack = [(1, 1)]  # (u, h)
    while stack:
        u, h = stack.pop()
        # walk heavy path
        x = u
        while x:
            head[x] = h
            cur += 1
            pos[x] = cur
            inv[cur] = x
            # push light children
            for v in g[x]:
                if v == parent[x] or v == heavy[x]:
                    continue
                stack.append((v, v))
            x = heavy[x]

    # LCA with binary lifting
    LOG = (N).bit_length()
    up = [[0]*(N+1) for _ in range(LOG)]
    for i in range(1, N+1):
        up[0][i] = parent[i]
    for k in range(1, LOG):
        prev = up[k-1]
        curu = up[k]
        for i in range(1, N+1):
            curu[i] = prev[prev[i]]

    def lca(a, b):
        if depth[a] < depth[b]:
            a, b = b, a
        diff = depth[a] - depth[b]
        bit = 0
        while diff:
            if diff & 1:
                a = up[bit][a]
            diff >>= 1
            bit += 1
        if a == b:
            return a
        for k in range(LOG-1, -1, -1):
            if up[k][a] != up[k][b]:
                a = up[k][a]
                b = up[k][b]
        return parent[a]

    seg = SegTree(N, Rbase)
    invR = modinv(Rbase)

    U, Q = map(int, input().split())

    updates = [tuple(map(int, input().split())) for _ in range(U)]
    queries = [tuple(map(int, input().split())) for _ in range(Q)]

    # quadratic in z:
    # (a1+z d1)(a2+z d2) = c0 + c1 z + c2 z^2
    # c2=d1*d2, c1=a1*d2+a2*d1, c0=a1*a2

    def apply_segment(q_is_R, Lpos, Rpos, d0, c0, c1, c2):
        # We need value at step i:
        # z = d0 + i (if q_is_R) OR z = d0 - i (if q_is_inv)
        # value = (c0 + c1 z + c2 z^2) * R^z
        # factor R^d0 and represent as (p0+p1*i+p2*i^2)*q^i
        d0m = d0 % MOD
        d02 = d0m * d0m % MOD

        base = pow(Rbase, d0, MOD)

        if q_is_R:
            # z=d0+i
            P2 = c2 % MOD
            P1 = (c1 + 2*c2*d0m) % MOD
            P0 = (c0 + c1*d0m + c2*d02) % MOD
            # multiply by R^d0
            p0 = P0 * base % MOD
            p1 = P1 * base % MOD
            p2 = P2 * base % MOD
            seg.range_add_seq(True, Lpos, Rpos, p0, p1, p2)
        else:
            # z=d0-i, exponent R^{d0}*(invR)^i
            P2 = c2 % MOD
            P1 = (-c1 - 2*c2*d0m) % MOD
            P0 = (c0 + c1*d0m + c2*d02) % MOD
            p0 = P0 * base % MOD
            p1 = P1 * base % MOD
            p2 = P2 * base % MOD
            seg.range_add_seq(False, Lpos, Rpos, p0, p1, p2)

    def path_update(A, B, a1, d1, a2, d2):
        A0 = a1 % MOD; D1 = d1 % MOD
        A2 = a2 % MOD; D2 = d2 % MOD
        c2 = D1 * D2 % MOD
        c1 = (A0 * D2 + A2 * D1) % MOD
        c0 = A0 * A2 % MOD

        L = lca(A, B)
        distAL = depth[A] - depth[L]

        # ---- A -> L part (includes L): z = depth[A]-depth[x] (decreasing along depth)
        u = A
        while head[u] != head[L]:
            h = head[u]
            Lpos = pos[h]
            Rpos = pos[u]
            d0 = depth[A] - depth[h]  # z at node h (left)
            apply_segment(False, Lpos, Rpos, d0, c0, c1, c2)  # base invR
            u = parent[h]
        # same head
        Lpos = pos[L]
        Rpos = pos[u]
        d0 = depth[A] - depth[L]
        apply_segment(False, Lpos, Rpos, d0, c0, c1, c2)

        # ---- L -> B part (exclude L): z = distAL + (depth[x]-depth[L]) increasing
        v = B
        segs = []
        while head[v] != head[L]:
            h = head[v]
            segs.append((pos[h], pos[v]))  # top->bottom in dfn
            v = parent[h]
        # same head: from L+1 to v if exists
        if v != L:
            segs.append((pos[L] + 1, pos[v]))

        # each segment is top->bottom increasing
        for lpos, rpos in segs:
            topnode = inv[lpos]
            d0 = distAL + (depth[topnode] - depth[L])
            apply_segment(True, lpos, rpos, d0, c0, c1, c2)

    def path_sum(u, v):
        res = 0
        while head[u] != head[v]:
            if depth[head[u]] < depth[head[v]]:
                u, v = v, u
            res = (res + seg.range_sum(pos[head[u]], pos[u])) % MOD
            u = parent[head[u]]
        if depth[u] > depth[v]:
            u, v = v, u
        res = (res + seg.range_sum(pos[u], pos[v])) % MOD
        return res

    # apply all updates
    for a1, d1, a2, d2, A, B in updates:
        path_update(A, B, a1, d1, a2, d2)

    # answer queries
    out = []
    for i, j in queries:
        out.append(str(path_sum(i, j) % MOD))
    sys.stdout.write("\n".join(out))

if __name__ == "__main__":
    solve()


# timesout